### ISSUE TITLE
Error when Export TRT model from the Quantized ONNX

### ISSUE THREAD

**Issue by chuong98** [NVIDIA/TensorRT-Model-Optimizer#16]
After successful quantizing and exporting ONNX models for ResNet18, using 2 different mode `int8` and `fp8`, I am trying to export these ONNX models to TRT, but no luck so far. It returns Error No support layers:?

```
[05/22/2024-03:49:37] [E] Error[9]: Skipping tactic 0xdb2ceb83bdb264c9 due to exception Assertion dims1.d[i] == 1 || dims2.d[i] == 1 || dims1.d[i] == dims2.d[i] failed. [64] cannot broadcast with [224]
[05/22/2024-03:49:37] [E] Error[9]: Skipping tactic 0xc37005486323d39b due to exception Assertion dims1.d[i] == 1 || dims2.d[i] == 1 || dims1.d[i] == dims2.d[i] failed. [64] cannot broadcast with [224]
[05/22/2024-03:49:37] [W] [TRT] Engine generation failed with backend strategy 2.
Error message: [optimizer.cpp::computeCosts::4048] Error Code 10: Internal Error (Could not find any implementation for node conv1.weight + /conv1/weight_quantizer/QuantizeLinear + /conv1/Conv.).
Skipping this backend strategy.
[05/22/2024-03:49:37] [E] Error[2]: [standardEngineBuilder.cpp::makeEngineFromSubGraph::1545] Error Code 2: Internal Error (Engine generation failed because all backend strategies failed.)
[05/22/2024-03:49:37] [E] Engine could not be created from network
[05/22/2024-03:49:37] [E] Building engine failed
[05/22/2024-03:49:37] [E] Failed to create engine from model or file.
[05/22/2024-03:49:37] [E] Engine set up failed
&&&& FAILED TensorRT.trtexec [TensorRT v9300] # trtexec --onnx=cache/int8/resnet18_quantized_int8.onnx --saveEngine=cache/int8/resnet18_quantized_int8.trt --fp16 --builderOptimizationLevel=4 --int8 --staticPlugins=/workspace/examples/plugins/bin/FP8Conv2DPlugin.so --staticPlugins=/workspace/examples/plugins/bin/groupNormPlugin.so --workspace=48000 --timingCacheFile=timing.cache
```

**Comment by chuong98** [comment]
Here is my command to export TRT:
```
fp8_plugin="/workspace/examples/plugins/bin/FP8Conv2DPlugin.so"
groupNorm_plugin="/workspace/examples/plugins/bin/groupNormPlugin.so"

# --minShapes=input:1x3x224x224 \
# --optShapes=input:16x3x224x224 \
# --maxShapes=input:32x3x224x224 \
# --shapes=input:1x3x224x224 \
export_params=$(cat <<-END
    --onnx=$weight_path \
    --saveEngine=$weight_dir/$new_filename \
    --fp16 --builderOptimizationLevel=4 \
    --int8 --staticPlugins=$fp8_plugin --staticPlugins=$groupNorm_plugin \
    --workspace=48000 --timingCacheFile=timing.cache
END
)

echo "Exporting TRT model ${new_filename}..."
trtexec $export_params
```

The onnx models can be downloaded from the Google Drive: https://drive.google.com/drive/folders/1kQEVN7FjLXD3XPF2Qrlr6gqNeiDVPup4?usp=sharing

**Comment by i-riyad** [comment]
> After successful quantizing and exporting ONNX models for ResNet18, using 2 different mode int8 and fp8

What command did you use to quantize the ONNX model?

**Comment by chuong98** [comment]
@i-riyad I attach the python script to quantize the model, following the documents
[quantize_model.txt](https://github.com/NVIDIA/TensorRT-Model-Optimizer/files/15406821/quantize_model.txt)
(Please rename the file from *.txt to*.py, GitHub does not allow uploading *.py)

Essentially, this is the snippet code to quantize:
```python
dataset_quant = create_dataset('imagenet', root='/data/imagenet', split='val', num_samples=512)
data_loader_quant = create_loader(dataset_val, 
                            input_size=(3, 224, 224), 
                            batch_size=args.batch_size, 
                            is_training=False, 
                            use_prefetcher=False)
# Define forward_loop. Please wrap the data loader in the forward_loop
def forward_loop(model):
    for batch in tqdm(data_loader_quant):
        input, target = batch
        input = input.to(device)
        model(input)
model_quant = mtq.quantize(model, config, forward_loop)
```

**Comment by chuong98** [comment]
It turns out that the ResNet has Conv and BN layers, the Quantizer insert Quantize and DeQuantize layers between Conv and BN, leads to fail. I got it works by:
- fuse the BN and Conv before Quantize.
- the default Quantize cfg doesn't work. Replace the config from the Diffuser then it works: `get_in8_config`, https://github.com/NVIDIA/TensorRT-Model-Optimizer/blob/main/diffusers/utils.py

**Comment by realAsma** [comment]
> It turns out that the ResNet has Conv and BN layers, the Quantizer insert Quantize and DeQuantize layers between Conv and BN, leads to fail. I got it works by:
> 
> * fuse the BN and Conv before Quantize.
> * the default Quantize cfg doesn't work. Replace the config from the Diffuser then it works: `get_in8_config`, https://github.com/NVIDIA/TensorRT-Model-Optimizer/blob/main/diffusers/utils.py

@chuong98 I infer that you used `INT8_SMOOTHQUANT_CFG`. This config is for INT8 quantization with smoothquant algorithm and is intended to be used in models with only `nn.Linear` layers. 

For quantizing CNN models to INT8, we need to use `INT8_DEFAULT_CFG`? 
Please see more about the quantization formats [here](https://nvidia.github.io/TensorRT-Model-Optimizer/reference/generated/modelopt.torch.quantization.config.html#quantization-formats).

If you are not happy with your current solution, would it be possible for you to try out `INT8_DEFAULT_CFG` and see if it works?

Thanks!!

**Comment by chuong98** [comment]
@realAsma Thanks for your help. Yes, I missed the Note: `The recommended configs given below are for LLM models. For CNN models, only INT8 quantization is supported. Please use quantization config INT8_DEFAULT_CFG for CNN models.` 

change the config `INT8_DEFAULT_CFG` works without manual fusing Conv-BN. Great.
Then, compare the `INT8_DEFAULT_CFG` to the option `trtexec --onnx= $MODEL --int8` that we commonly use, what are their difference? Are they the same? Thank you.

**Comment by chuong98** [comment]
My experiment with TensorRT Optimize Model yield worse results than the default TensorRT `trtexec`:
![image](https://github.com/NVIDIA/TensorRT-Model-Optimizer/assets/15174756/72c55b10-9622-4899-ba15-18aaada0cd6a)

**Comment by jingyu-ml** [comment]
I noticed you used `generate_fp8_scale` for the fp8 ResNet. This function is necessary because `torch.onnx.export` currently doesn't support FP8 export for Conv layers. To work around this, we borrowed the int8 export logic. Then, at this [link](https://github.com/NVIDIA/TensorRT-Model-Optimizer/blob/main/diffusers/onnx_utils/sdxl_fp8_graphsurgeon.py), we manually change all the int8 QDQ operations to FP8 in onnx graph.

We will soon release FP8 support in TensorRT out-of-the-box. Once it's available, you can use `generate_fp8_scale` to convert the model to a TensorRT engine without needing a graph surgeon. Additionally, once we verify that `torch.dynamo_export` works for FP8 Conv layers, we can try exporting without generate_fp8_scale.

**Comment by chuong98** [comment]
Can you provide an example to quantize a ViT model, for example Swin?

**Comment by realAsma** [comment]
@chuong98 
My apologies on the performance degradation. `modelopt.torch.quantization` speedup analysis have been focused on  LLMs (deployed via TensorRT-LLM) and diffusion models (deployed via TRT).

It is quite possible for quantized models other than the LLMs or diffusion models to be slower than TRT's un-quantized baseline models. 
In this case, we recommend exporting the Pytorch model to ONNX first and then quantize the ONNX graph via [modelopt.onnx.quantization](https://nvidia.github.io/TensorRT-Model-Optimizer/guides/_onnx_quantization.html).  Please see examples for quantizing ONNX graph [here](https://github.com/NVIDIA/TensorRT-Model-Optimizer/tree/main/onnx_ptq).

Could you please try out quantizing the ONNX graph using `modelopt.onnx.quantization` instead?

**Comment by korkland** [comment]
> My experiment with TensorRT Optimize Model yield worse results than the default TensorRT `trtexec`: ![image](https://private-user-images.githubusercontent.com/15174756/333370585-72c55b10-9622-4899-ba15-18aaada0cd6a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjE4OTQ5NzIsIm5iZiI6MTcyMTg5NDY3MiwicGF0aCI6Ii8xNTE3NDc1Ni8zMzMzNzA1ODUtNzJjNTViMTAtOTYyMi00ODk5LWJhMTUtMThhYWFkYTBjZDZhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MjUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzI1VDA4MDQzMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTYxNDIxYWEzMmQ0NjE5MzU0OTNiODIyM2RmNDMzNGY4Njc1YTRiYzliYmJlYjg5YjM1NTkyN2NhMDNkMjA5OWYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.Ch783mggSv1S3Kb6C20rRoarnIv9MjzpUb9dsOWyHvA)

I'm experiencing the same degradation with int8 vs. fp16 in RetinaNet, using the modelopt PyTorch API. In your table, you compared fused BN vs. non-fused BN. Did you use the smooth config there or the int8 default config?

Have you found a better approach for quantization that gives you better latency with more control over the quantization process (not the implicit one, which, according to the table, seems to bring significant improvement in latency)?

**Comment by WhiteTeaDragon** [comment]
> I noticed you used `generate_fp8_scale` for the fp8 ResNet. This function is necessary because `torch.onnx.export` currently doesn't support FP8 export for Conv layers. To work around this, we borrowed the int8 export logic. Then, at this [link](https://github.com/NVIDIA/TensorRT-Model-Optimizer/blob/main/diffusers/onnx_utils/sdxl_fp8_graphsurgeon.py), we manually change all the int8 QDQ operations to FP8 in onnx graph.
> 
> We will soon release FP8 support in TensorRT out-of-the-box. Once it's available, you can use `generate_fp8_scale` to convert the model to a TensorRT engine without needing a graph surgeon. Additionally, once we verify that `torch.dynamo_export` works for FP8 Conv layers, we can try exporting without generate_fp8_scale.

@jingyu-ml Hello! Thank you very much for explaining what goes wrong in the process of exporting the model! However, I have looked at the example code from diffusers and haven't fully understood it. More specifically, I do not understand why it is necessary to cast inputs and outputs into fp16 in this [line](https://github.com/NVIDIA/TensorRT-Model-Optimizer/blob/f8114cb37f5fcdb549467f87031d7616967b984f/examples/diffusers/quantization/onnx_utils/export.py#L253) and why one needs to convert float to float16 in this [line](https://github.com/NVIDIA/TensorRT-Model-Optimizer/blob/f8114cb37f5fcdb549467f87031d7616967b984f/examples/diffusers/quantization/onnx_utils/export.py#L259). Why it is done only if the desired format is fp8? Why is fp16 instead of fp8 used in these two lines?

**Comment by jingyu-ml** [comment]
@WhiteTeaDragon 

1. Cast the input and output to FP16/BF16

This issue exists in TRT 9.x, where TensorRT struggles with casting models correctly. However, in TRT 10.x, this is no longer necessary, so you don’t need to cast the input and output to FP16/BF16.

1. Convert float to float16

This was a workaround for an older ModelOPT ONNX export (versions around 0.1x.0). We observed that normalization weights were always in float32 within the ONNX and TensorRT engine, even when the original model used BF16/FP16. This bug has since been fixed, ensuring all weights remain in their original precision (BF16/FP16). The conversion code remains as a precaution, but you can remove it if you want a cleaner implementation.

The first point is specific to FP8, while the second is not. However, since these issues have been resolved in the latest TRT and ModelOPT versions, you can ignore them and remove the related code from your script.

Overall, if you're interested in using Conv for FP8, always start with the latest TensorRT version—currently, I believe it's 10.8. Before exporting to ONNX, convert the quantizers to INT8 using a process like this:

```python
            module.input_quantizer._num_bits = 8
            module.weight_quantizer._num_bits = 8
            module.input_quantizer._amax = module.input_quantizer._amax * (127 / 448.0)
            module.weight_quantizer._amax = module.weight_quantizer._amax * (127 / 448.0)

```

After obtaining the ONNX model, modify the QDQ nodes for the Conv layer directly in the ONNX graph. You can refer to this script:
https://github.com/NVIDIA/TensorRT-Model-Optimizer/blob/release/0.23.0/examples/diffusers/quantization/onnx_utils/fp8_onnx_graphsurgeon.py#L105

If you're looking to deploy the DiT model, no major adjustments are needed. You can refer to the Flux example for guidance. The RoPE issue has also been resolved in TensorRT 10.x, so you can safely ignore it.

https://github.com/NVIDIA/TensorRT-Model-Optimizer/blob/main/examples/diffusers/quantization/onnx_utils/export.py#L266C13-L266C42

Please let us know if you run into anther issue.

**Comment by WhiteTeaDragon** [comment]
@jingyu-ml 

Thank you very much for your answer! I have successfully quantized torch model and exported it to onnx. I also have a small question about example command for trtexec named 'For FLUX-Dev FP8' from here: https://github.com/NVIDIA/TensorRT-Model-Optimizer/blob/main/examples/diffusers/quantization/README.md

Why do you use stronglyTyped together with `--fp8 --bf16`? In the [docs](https://docs.nvidia.com/deeplearning/tensorrt/10.8.0/inference-library/advanced.html#strongly-typed-networks), it is written that when building the engine with the flag `--stronglyTyped`, other precision settings are ignored.

In my experiments, I have the following hierarchy of the final trt-model efficiency depending on convertation settings:
`quantized onnx model + flag --stronglyTyped` is slower than `quantized onnx model + flag --fp8`, which is still slower than `quantized onnx model + flags --fp8 --fp16`.

Moreover, preliminary quantization of the model does not affect performance, except for the case with flag `--stronglyTyped`. In this case, quantized model actually works slower than non-quantized model.

Is this the expected behaviour? Does quantization of torch model with this library only help with accuracy of the final model, not the performance?

**Comment by jingyu-ml** [comment]
@WhiteTeaDragon 

Sorry for the late reply. I don’t have an answer for trtexec at the moment, but I’ll check internally and get back to you.

In the meantime, could you share a dummy ONNX model, along with the trtexec command you’re using, the TensorRT version, and the GPU type?

**Comment by WhiteTeaDragon** [comment]
@jingyu-ml 

Sorry for the late reply as well. I have figured out that fp8 is not supported by TensorRT for some convolutional layers in my model (large padding and kernel size, see rules here: https://docs.nvidia.com/deeplearning/tensorrt/archives/tensorrt-1060/release-notes/index.html). While I still think that fp8 should provide at least some speedup for the layers which can be quantized (and it is the majority of the layers in the model), it is probably not a very good example.
